CRM/Member/Form/MembershipRenewal.php

```diff
@@ -339,6 +339,12 @@ public function buildQuickForm() {
 
     $this->applyFilter('__ALL__', 'trim');
 
+    // CRM-20571
+    // Get the Join Date from Membership info as it is not available in the Renew form
+    $joinDate = CRM_Core_DAO::getFieldValue('CRM_Member_DAO_Membership', $this->_id, 'join_date');
+    // Save its value in a hidden field
+    $this->addElement('hidden', 'join_date', $joinDate);
```

Comment:
I think you have a good explanation and accepted answer from Andrew. Not sure if they could be done directly in buildForm() within a single line, but the current code looks fine and more straightforward to me. But if you want to try, I think you can do something like - https://pear.php.net/manual/en/package.html.html-quickform.html-quickform.addrule.php ?

For options accepted by `addRule()`,  these are the registered rules within quickform https://pear.php.net/manual/en/package.html.html-quickform.html-quickform.getregisteredrules.php. You can directly see the list of them at https://github.com/civicrm/civicrm-packages/blob/master/HTML/QuickForm.php#L75-L89.

Also civicrm registers some rules explicitly available on all forms - see https://github.com/civicrm/civicrm-core/blob/master/CRM/Core/Form.php#L288. So you can use any of them.